### PR TITLE
Fullscreen setting GUI option

### PR DESCRIPTION
**Pull request by neoascetic**

### CODE REVIEW

**Comment by neoascetic** [comment]
It should be disabled on Android probably.

**Comment by Zeno-** [comment]
@neoascetic yes, this needs to be disabled on Android -- you are correct

**Comment by neoascetic** [comment]
@Zeno- done!

**Comment by kilbith** [comment]
As the last designer of that tab, I recommend you to extend downward the left part and move this checkbox inside.

**Comment by neoascetic** [comment]
I thought central column is for rendering details (for which fullscreen setting belongs to) and left and right columns are for different fancy effects, that's why I put the settings there.

**Comment by est31** [comment]
Disable it [like done here](https://github.com/neoascetic/minetest/blob/fullscreen-setting/builtin/mainmenu/init.lua#L44).

File: builtin/mainmenu/tab_settings.lua [hunk 1]
```diff
@@ -203,13 +203,15 @@ local function formspec(tabview, name, tabdata)
...
-		"box[7.75,0;4,4;#999999]" ..
+		"box[7.75,0;4,3.9;#999999]" ..
 		"checkbox[8,0;cb_shaders;".. fgettext("Shaders") .. ";"
 				.. dump(core.setting_getbool("enable_shaders")) .. "]"
 
 	if PLATFORM ~= "Android" then
 		tab_string = tab_string ..
-		"button[8,4.75;3.75,0.5;btn_change_keys;".. fgettext("Change keys") .. "]"
+		"button[8,4.75;3.75,0.5;btn_change_keys;".. fgettext("Change keys") .. "]" ..
+		"checkbox[3.85,3.15;cb_fullscreen;".. fgettext("Fullscreen (requires restart)") .. ";" ..
```
Comment (neoascetic) [inline builtin/mainmenu/tab_settings.lua]: @est31 I've already disabled it here.

**Comment by C1ffisme** [comment]
Why do you need to restart for the fullscreen to take effect? You can already press F11 and you need no restart.

**Comment by neoascetic** [comment]
@C1ffisme this doesn't work for me. And I cannot see keybinding for this in main menu. Where did you get this info?

**Comment by C1ffisme** [comment]
@neoascetic I dunno, I saw it one time on an issue discussion. Maybe it's just cause I run kubuntu. It works on my current version.

**Comment by NathanSalapat** [comment]
I too can verify F11 doesn't work for me. Running Linux Mint 17.1 Cinnamon edition.

**Comment by C1ffisme** [comment]
I am running an older version. Which may have something to do with it.

**Comment by est31** [comment]
What version do you use @C1ffisme ?

**Comment by C1ffisme** [comment]
I'm using 0.4.12-dev, but I'm not using the latest build. I might wait out until schematic problems are fixed.

I'm not really sure how to get clearer than 0.4.12-dev. Is there a way to get the date?

**Comment by est31** [comment]
You'll need to try a tool specific to the way how you obtained minetest. For the daily ppa, try `dpkg -s minetest`.

**Comment by C1ffisme** [comment]
Ah, maybe later...

**Comment by rubenwardy** [comment]
If F11 does work, it will be because of the operating system / window manager, and not because of Minetest.

**Comment by kilbith** [comment]
I've tested this PR. That fullscreen option is a pretty bad idea and does not work the same way than F11.

Irrlicht only support fullscreen as a window creation option, this means that if you enable fullscreen in .conf, the display resolution of the client is kept but enlarged on the whole screen (the result is awful).

**Comment by neoascetic** [comment]
@kilbith Why it is bad idea? I even didn't know that there is such option in configuration file! In my opinion, it is better for end-user to be able to do this via option instead of configuration file editing.

**Comment by kilbith** [comment]
@neoascetic Your PR **adds** the fullscreen setting in minetest.conf

**Comment by neoascetic** [comment]
@kilbith you are not right.

**Comment by Zeno-** [comment]
F11 is a window manager / desktop environment thing.

**Comment by kilbith** [comment]
> You are not right.

You should open and read your minetest.conf before saying bullshit...

**Comment by neoascetic** [comment]
@kilbith [this](https://github.com/minetest/minetest/blob/master/src/client/clientlauncher.cpp#L501) is not my code. Yes, click on that checkbox will **set** this option in config file, but the PR doesn't **introduce** this option.

**Comment by kilbith** [comment]
@neoascetic This is what I meant...

**Comment by neoascetic** [comment]
Anyway, I have no issues with fullscreen enabled. @kilbith could you provide a screenshot with "awful" result?

screenshot_1367696600

**Comment by kilbith** [comment]
@neoascetic Delete the screenH and screenW in your .conf, and enable fullscreen to true, and then you'll get my point.

**Comment by neoascetic** [comment]
@kilbith I done this. Nothing changed. Those options only affects initial window width/height in windowed mode for me.

**Comment by kilbith** [comment]
@neoascetic Probably because you're on Mac, but on Linux it change the resolution of the system to the  resolution of the Minetest client.

**Comment by neoascetic** [comment]
Yes, problem you've described also appears on windows, but not on OSX. Hm.

**Comment by neoascetic** [comment]
It would be better to have this **with** (not **instead of**) #2555

**Comment by neoascetic** [comment]
> @kilbith, [here](https://github.com/minetest/minetest/pull/2555#issuecomment-86096570)
> I'd have prefered to discuss of that on your PR. To answer you : no, even with them. If you set something lower than your max. screen resolution and enable fullscreen = true, the MT client resolution overwrite the system resolution and/or the window manager is screwed up.

Could you please explain "screwed up"? On windows, I just have large black borders around the screen with the game. But this is common for any game that allows you to change resolution and window mode. User should know what he is doing.

**Comment by kilbith** [comment]
@neoascetic 

On Linux, tested with : 

```
screenW = 800 (or any other width)
screenH = 600 (or any other height)
fullscreen = true
```

or just with `fullscreen = true`

The MT client displaying takes place of the desktop environment, whether in "window" mode (without borders / window managering), or in full screen in overwritting the system resolution by the MT resolution. A screenshot was not possible, and the cursor couldn't go beyond the MT "area"...

sam_0076

**Comment by Zeno-** [comment]
@kilbith 

Fullscreen works (and will scale to fit the entire screen) **if** the resolution is any of the supported videomodes listed by minetest --videomodes (or my new Lua function).  You only get the screen looking like that in your photo if the fullscreen resolution is not a supported resolution.

**Comment by Pilcrow182** [comment]
I think the problem is in the fullscreen minetest.conf option itself: when fullscreen is enabled and no screenW or screenH are present, it appears minetest defaults to a resolution of 800x600. Some screens and video cards (including, apparently, kilbith's) don't support 800x600. Instead of defaulting to a static resolution of 800x600, minetest should check the native resolution and default to that.

I can confirm, however, that the fullscreen option works in *most* linux and windows systems, even without screenW and screenH present. In fact, for *all* of the computers I've tried (around 11 of them) the problem in kilbith's screenshot *only* appears when screenW and screenH are specifically set to an unsupported resolution. When they're not present at all, it runs at 800x600 (which is a bit low-res and ugly, but still scales correctly)...

I think this PR should be merged, as there seems to be absolutely nothing wrong with neoascetic's code (and it's a needed change; I've always wondered why a fullscreen GUI option didn't exist already), but it might be a good idea to touch up the resolution code itself first.

**Pull request closed without merging**